### PR TITLE
rustc: Link LLVM directly into rustc again (take two)

### PR DESCRIPTION
**Pull request by Aaron1011**
This is a continuation of PR https://github.com/rust-lang/rust/pull/65703

### CODE REVIEW

**Comment by Aaron1011** [comment]
@alexcrichton: I can't reproduce the CI failure locally. Could we run a try-build to see if it's still an issue?

**Comment by bjorn3** [comment]
When not merging this PR, it would be possible to rewrite https://github.com/getditto/rust-bitcode to only recompile the codegen backend, instead of a whole rust compiler by using `-Zcodegen-backend`.

**Comment by alexcrichton** [comment]
I'd recommend enabling the try builder as part of this PR itself, you can edit the configuration for builders on PRs so that way you can test through a PR if you're unable to test it locally.

**Comment by Aaron1011** [comment]
I've managed to reproduce the issue locally via `TF_BUILD=True ./src/ci/docker/run.sh dist-x86_64-linux`

**Comment by Aaron1011** [comment]
@alexcrichton: I believe I've fixed the issue. It didn't show up locally because I was statically linking LLVM, instead of dynamicly linking against `libLLVM.so`

**Comment by Aaron1011** [comment]
It looks like this actually passed, but failed due to the weird.CI change.

@alexcrichton: I think this is ready to merge

**Comment by alexcrichton** [comment]
Ok great! Want to back out the CI changes?

Additionally, there's another call to `dist::maybe_install_llvm_dylib`, but should this one be the source of truth? Should the one in `Assemble` get removed?

**Comment by Aaron1011** [comment]
@alexcrichton: The existing call to `maybe_install_llvm_dylib` is a little different - it is run during `Assemble`, and is responsible for copying LLVM into the *next* stage - e.g. when we are copying `stage0` artifacts into the `stage` dir. This ensures that LLVM exists when we run the build stage (e.g. start running `stage` rustc).

The new call to `maybe_install_llvm_dylib` is specifically for `stage0` - it is responsible for installing LLVM to the *existing* `stage0`, so that tools building against `stage0` can see it. This is the wrong behavior for subsequent stages - we don't want to copy LLVM into the *current* sysroot directory - we want to to be copied into the built `n + 1` sysroot, so it can used by that compiler.

**Comment by Aaron1011** [comment]
I've added a comment to the existing call explaining what it does.

**Comment by michaelwoerister** [comment]
I just noticed that this might interact with how we do ThinLTO on the LLVM code which gives us something like a 10% performance boost. Is that still possible now?

**Comment by Aaron1011** [comment]
@michaelwoerister: I'm not quite sure what you mean - are you saying we currently get a 10% performance boost, or that this PR might give us a 10% performance boost?

This PR does not affect how LLVM itself is built (e.g. whether it's statically or dynamically linked into `librustc_codegen_llvm`. However, it does open up the possibility of doing cross-language LTO with both Rust and LLVM (since `librustc_codegen_llvm` is now built as a normal crate).

**Comment by nikic** [comment]
We should probably at least to a perf run, it looks like neither this nor the previous PR had one.

@bors try @rust-timer queue

**Comment by alexcrichton** [comment]
@michaelwoerister I don't believe that this changes anything as @Aaron1011 said, we're simply statically linking librustc_codegen_llvm into rustc instead of dynamically linking it. For release builds on Linux/Mac LLVM is still dynamically linked. On Windows LLVM will now be statically linked into the single rustc executable.

We already do ThinLTO on LLVM itself when producing the LLVM dynamic library, but this is only on Linux. We don't do ThinLTO with all of LLVM and Rust code currently.

r=me with the new comment, thanks @Aaron1011! Seems prudent to wait for confirmation with a perf run though that there's no regressions.

**Comment by michaelwoerister** [comment]
As long as LLVM itself is still a dylib ThinLTO shouldn't be affected. Thanks for clarifying that, @alexcrichton!

**Comment by nikic** [comment]
Perf looks mildly positive for small crates, no surprises. This needs a rebase for Cargo.lock though.

**Comment by Aaron1011** [comment]
@nikic @alexcrichton: Rebased

**Comment by Aaron1011** [comment]
@alexcrichton: The new build configuration seems to have exposed a weird implicit dependency of rustllvm on rustc_codegen_llvm. I've pushed a commit that should fix the issue.

**Comment by alexcrichton** [comment]
@bors: r+

Nice catch!

File: src/bootstrap/compile.rs [hunk 1]
```diff
@@ -501,6 +538,49 @@ pub fn rustc_cargo_env(builder: &Builder<'_>, cargo: &mut Cargo) {
...
+        }
+        // Some LLVM linker flags (-L and -l) may be needed to link librustc_llvm.
+        if let Some(ref s) = builder.config.llvm_ldflags {
+            cargo.env("LLVM_LINKER_FLAGS", s);
+        }
+        // Building with a static libstdc++ is only supported on linux right now,
+        // not for MSVC or macOS
+        if builder.config.llvm_static_stdcpp &&
+           !target.contains("freebsd") &&
+           !target.contains("windows") &&
```
Comment (mati865) [inline src/bootstrap/compile.rs]: This line changed `msvc` to `windows` breaking windows-gnu target.

**Pull request merged**